keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -117,59 +116,44 @@ serves the resourceVersion="0" list requests from reflectors today.
 
 Guard this by a `WatchCacheConsistentReads` feature gate.
 
-#### Use WithProgressNotify to enable automatic watch updates
+This alternative requires using `WatchProgressRequest` which is only available in etcd 3.4+, and so would
+require we make the kube-apiserver aware of etcd's minor version, which is described in more detail later.
 
-Create etcd watches with `WithProgressNotify` enabled (available in all etcd 3.x versions).
+<<[UNRESOLVED @serathius]>>
+Propose a method to make kube-apiserver aware of etcd version:
+* Flag providing etcd version to kube-apiserver `--storage-backend=etcd3.4`
+* Flag to enable/disable the feature `--allow-using-progress-notify`
+* Have kube-apiserver check cluster version in etcd `/version` endpoint. 
+  Retry the check logic if `WatchProgressRequest` fails.
+<<[/UNRESOLVED]>>
 
-When `WithProgressNotify` is enabled on an etcd watch, etcd sends progress
-events to the watch automatically. By default etcd sends progress events every
-10 minutes, which is not frequent enough to be useful for our needs, so we will
-modify etcd to send them more frequently.
+#### Use RequestProgress to enable automatic watch updates
 
-When an consistent LIST request is received and the watch cache is enabled:
+When a consistent LIST request is received and the watch cache is enabled:
 
-- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Get the current revision from etcd for the resource type being served. Use 
+  an etcd range request against a known empty range with limit=1 to ensure returned revision is strongly consistent.
+- If the cache already has the current revision, serve the request from cache. If not,
+  - Send a `WatchProgressRequest` to etcd on the watch channel that the watch cache is consuming.
 - Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly for the watch to catch up to the current revision.
 - If the block times out, the request will result in rejection. (see "What if the watch cache is stale?" section for details)
 
-To get the revsion we have some options:
-
-- Use an etcd range request with `WithCount` enabled so etcd return only a count and revision
-- Use an etcd range request against a known empty range with limit=1 as an additional guard (since etcd does not allow for limit=0)
-
 Consistent GET requests will continue to be served directly from etcd. We will
 only serve consistent LIST requests from cache.
 
-Important: We are planning to set the progress notify interval to 250ms, which will introduce up to 250ms latency to consistent LIST requests.
-
-Optional: For some (but not all) of the etcd progress watch events, also create a
+For the etcd progress notification, we will create a
```

Comment:
I want to revisit this - I don't think it's the right decision.

With other efforts, we enabled periodic progress-notify events every 5s (at least in some environments). The explicitly requested-ones can make them even more frequent.
I don't want to send bookmark to every watcher in the system (potentially hundreds of thousands) every 5s or even more frequently - that has significant performance implication.